Query range fields by doc values when they are expected to be more efficient than points

martijnvg opened the pull request:
* Enable doc values for range fields by default.
* Store ranges in a binary format that support multi field fields.
* Added BinaryDocValuesRangeQuery that can query ranges that have been encoded into a binary doc values field.
* Wrap range queries on a range field in IndexOrDocValuesQuery query.

Closes #24314

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -467,6 +480,50 @@ public InetAddress nextUp(Object value) {
+            public BytesRef encodeRanges(Set<Range> ranges) throws IOException {
+                final byte[] encoded = new byte[ByteUtils.MAX_BYTES_VLONG + (16 * 2) * ranges.size()];
+                ByteArrayDataOutput out = new ByteArrayDataOutput(encoded);
+                out.writeVInt(ranges.size());
```
since we use a fixed length, maybe we do not need to encode the number of values?

jpountz replied:
oh I see it is to be consistent with other types which use vlong

martijnvg commented:
> I think we should design it in such a way that queries can work directly on the encoded representation?

The idea here is that we can save some space by encoding the ranges as two vlongs. Downside is that we can't work on the encoded representation. Do you think that the vlong encoding isn't worth the effort here? Personally I'm ok with changing the encoding here, so that we don't to encode at query time, but just wondering if we should make a tradeoff between space or query speed.

jpountz commented:
You convinced me that we need to perform some sort of compression for numeric values. However I think we should use an encoding that doesn't require the query to know whether the field stores longs or ip addresses?

jpountz commented on core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java:
```diff
@@ -0,0 +1,181 @@
+        }
+        if (includeTo) {
+            to = rangeType.nextUp(to);
+        }
```
This looks a bit backward to me, we would usually rather do
```
if (includeFrom == false) {
    from = rangeType.nextUp(from);
}
```
Do I miss something?

jpountz commented on core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java:
```diff
@@ -0,0 +1,181 @@
+            ByteArrayDataInput in = new ByteArrayDataInput(encodedRanges.bytes, encodedRanges.offset, encodedRanges.length);
+            int numRanges = in.readVInt();
+            for (int i = 0; i < numRanges; i++) {
+                int length = in.readVInt();
```
the first bytes are not a vint when we encode eg. longs? So I don't think it works?

jpountz commented on core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java:
```diff
@@ -0,0 +1,181 @@
+                length = in.readVInt();
+                bytes = new byte[length];
+                in.readBytes(bytes, 0, length);
+                BytesRef otherTo = new BytesRef(bytes);
```
We should try to avoid ByteArrayDataInput and BytesRef allocations since this method may be called in a tight loop.

martijnvg replied:
I can remove the ByteRef instance creation and use just plain byte[]. 
The reason I used `ByteArrayDataInput` is because of its `readVInt()` and `readBytes()` methods. Are there static alternatives that I can use? Otherwise maybe it is sufficient enough if I keep a spare instance around that I keep reusing?

jpountz commented on core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java:
```diff
@@ -0,0 +1,181 @@
+                        // does not disjoint AND not within:
+                        result =  (from.compareTo(otherTo) > 0 || to.compareTo(otherFrom) < 0) == false &&
+                            (from.compareTo(otherFrom) <= 0 && to.compareTo(otherTo) >= 0) == false;
+                        break;
```
should we put a method on the queryType enum instead of this switch?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -759,21 +874,102 @@ public Object parseTo(RangeFieldType fieldType, XContentParser parser, boolean c
+            }
+            return encoded;
+        }
+
```
I'm wondering that putting these helpers into a dedicated class would make things easier to read?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -759,21 +874,102 @@ public Object parseTo(RangeFieldType fieldType, XContentParser parser, boolean c
+                if (sign == 1) {
+                    encoded[encoded.length - 1 - b] = (byte) (l >>> (8 * b));
+                } else if (sign == 0) {
+                    encoded[encoded.length - 1 - b] = (byte) (0xFF - (l >>> (8 * b)));
```
I think we miss a mask here: `encoded[encoded.length - 1 - b] = (byte) (0xFF - ((l >>> (8 * b)) & 0xFF));`

jpountz commented on core/src/test/java/org/apache/lucene/queries/LongRandomBinaryDocValuesRangeQueryTests.java:
```diff
@@ -0,0 +1,121 @@
+        }
+        long max = Long.MAX_VALUE / 2;
+        return (max + max) * random().nextLong() - max;
+    }
```
We had similar random number generation in eg. `TestIntRangeFieldQueries` but this proved inefficient at finding bugs (eg. it makes it very unlikely to get twice the same value in the same test) so we changed it in https://issues.apache.org/jira/browse/LUCENE-7847. Maybe we should do something similar here (as well as for doubles, ip addresses, etc.)

jpountz commented on core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java:
```diff
@@ -0,0 +1,181 @@
+    private final QueryType queryType;
+    private final NormalizedRange expectedRange;
+
+    public BinaryDocValuesRangeQuery(String fieldName, RangeFieldMapper.RangeType rangeType, QueryType queryType, Object from, Object to,
```
Can we avoid passing the range type? I think it would be more robust if it does not need to know about the type of values that are stored and works directly on the encoded data?

martijnvg commented:
@jpountz Thanks for reviewing. I've updated the PR.

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java:
```diff
@@ -0,0 +1,91 @@
+            l = Double.doubleToRawLongBits(d);
+            sign = 1; // means positive
+        }
+        return encode(l, sign);
```
doubles should probably use a different encoding, but we can work on it as a follow-up

jpountz replied:
Also I think we should specialize floats too

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java:
```diff
@@ -0,0 +1,91 @@
+        return new BytesRef(encoded, 0, out.getPosition());
+    }
+
+    static byte[] encode(long l) {
```
Also I think we need some unit tests for that method in order to ensure it actually preserves ordering?

martijnvg commented:
@jpountz I've updated the pr.

>  I'm wondering whether we should sort the ranges before encoding them as it might allow for optimizations in the multi-valued case eg. if the max value of the query is less than the min value of the first encoded range.

I've added the sorting, but I think it will only make sense when there multiple adjacent ranges and when ranges are overlapping it wouldn't help that much?

> It would be nice to have some benchmarks as well in order to make sure that this actually performs better than point-based ranges when the range does not drive iteration (it might make more sense to do it after LUCENE-7828 is merged). I can help with that.

Thanks! I guess we will need a dataset that when benchmarked with a specific query forces the range query to be executed in random access mode.

jpountz commented:
> I've added the sorting, but I think it will only make sense when there multiple adjacent ranges and when ranges are overlapping it wouldn't help that much?

Right I don't think this is a game changer for queries in general. However it might be the case for [aggregations](https://github.com/elastic/elasticsearch/issues/23182)? For instance, say you want to run an histogram aggregation on a range field, it would be easier with sorted ranges, since you can easily track which buckets have already been incremented and which buckets haven't?

jpountz commented:
I merged LUCENE-7828 yesterday so that we can better benchmark this change, but upgrading to a new Lucene 7 snapshot is currently blocked on #25028 since the postings highlighter has been removed from Lucene master. It should hopefully be merged in the next few days.

martijnvg commented:
This are the 99th percentile latency results between master (baseline) and this pr (contender) of the [new benchmark](https://github.com/elastic/rally-tracks/pull/30) added to rally:

|                        Metric |       Operation |   Baseline |   Contender |               Diff |   Unit |
| --------------------------- | ------------------  | ------------- | ---------------- |----------------- | ------- |
|       99th percentile latency |                                range_field_big_range |    4961.66 |     4643.11 | -318.556 |     ms |
|       99th percentile latency |                              range_field_small_range |    37.6339 |     29.3083 | -8.32565 |     ms |
|       99th percentile latency |   range_field_conjunction_big_range_small_term_query |     11.279 |     8.58807 |  -2.6909 |     ms |
|       99th percentile latency | range_field_conjunction_small_range_small_term_query |    6.15329 |     8.36214 |  2.20885 |     ms |
|       99th percentile latency |   range_field_conjunction_small_range_big_term_query |    2977.32 |     3478.42 |  501.095 |     ms |
|       99th percentile latency |     range_field_conjunction_big_range_big_term_query |    22320.1 |     21216.1 | -1104.07 |     ms |
|       99th percentile latency | range_field_disjunction_small_range_small_term_query |    1438.64 |     45.5729 | -1393.07 |     ms |
|       99th percentile latency |   range_field_disjunction_big_range_small_term_query |    8843.91 |      5680.8 | -3163.11 |     ms |

martijnvg commented:
@jpountz I've updated the pr.
